core/stat/base/leap_array.go

```diff
@@ -0,0 +1,203 @@
+	return ret
+}
+
+func (aa *atomicWindowWrapArray) elementOffset(idx int) uintptr {
+	if idx >= aa.length && idx < 0 {
+		panic(fmt.Sprintf("The index (%d) is out of bounds, length is %d.\n", idx, aa.length))
+	}
+	base := aa.base
+	return base + uintptr(idx*PtrSize)
+}
+
+func (aa *atomicWindowWrapArray) get(idx int) *windowWrap {
+	// get the secondary pointer of windowWrap, the pointer {offset} to the aa.data[idx]
+	offset := (**windowWrap)(unsafe.Pointer(aa.elementOffset(idx)))
+	wwPtr := (*unsafe.Pointer)(unsafe.Pointer(offset))
+	return (*windowWrap)(atomic.LoadPointer(wwPtr))
+}
+
+func (aa *atomicWindowWrapArray) compareAndSet(idx int, except, update *windowWrap) bool {
+	// get the secondary pointer of windowWrap, the pointer {offset} to the aa.data[idx]
+	offset := (**windowWrap)(unsafe.Pointer(aa.elementOffset(idx)))
+	wwPtr := (*unsafe.Pointer)(unsafe.Pointer(offset))
+	// update secondary pointer
+	return atomic.CompareAndSwapPointer(wwPtr, unsafe.Pointer(except), unsafe.Pointer(update))
+}
+
+// The windowWrap leap array,
+// sampleCount represent the number of windowWrap
+// intervalInMs represent the interval of leapArray.
+// For example, windowLengthInMs is 500ms, intervalInMs is 1min, so sampleCount is 120.
+type leapArray struct {
+	windowLengthInMs uint32
+	sampleCount      uint32
+	intervalInMs     uint32
+	array            *atomicWindowWrapArray
+	// update lock
+	updateLock triableMutex
+}
+
+func (la *leapArray) currentWindow(bg bucketGenerator) (*windowWrap, error) {
+	return la.currentWindowWithTime(util.CurrentTimeMillis(), bg)
+}
+
+func (la *leapArray) currentWindowWithTime(timeMillis uint64, bg bucketGenerator) (*windowWrap, error) {
+	if timeMillis < 0 {
+		return nil, errors.New("Current time is less than 0 \n")
+	}
+
+	idx := la.calculateTimeIdx(timeMillis)
+	windowStart := calculateStartTime(timeMillis, la.windowLengthInMs)
+
+	for { //spin to get the current windowWrap
+		old := la.array.get(idx)
+		if old == nil {
+			// because la.array.data had initiated when new la.array
+			// theoretically, here is not reachable
+			newWrap := &windowWrap{
+				windowStart: windowStart,
+				value:       bg.newEmptyBucket(),
+			}
+			if la.array.compareAndSet(idx, nil, newWrap) {
+				return newWrap, nil
+			} else {
+				runtime.Gosched()
+			}
+		} else if windowStart == atomic.LoadUint64(&old.windowStart) {
+			return old, nil
+		} else if windowStart > atomic.LoadUint64(&old.windowStart) {
+			// current time has been next cycle of leapArray and leapArray dont't count in last cycle.
+			// reset windowWrap
+			if la.updateLock.TryLock() {
+				old = bg.resetWindowTo(old, windowStart)
+				la.updateLock.Unlock()
+				return old, nil
+			} else {
+				runtime.Gosched()
+			}
+		} else if windowStart < old.windowStart {
+			// should not go through here,
+			return nil, errors.New(fmt.Sprintf("Provided time timeMillis=%d is already behind old.windowStart=%d \n", windowStart, old.windowStart))
```

Comment:
This control branch might be reserved for some special cases (e.g. when occupying "future" buckets).

Reply:
ic, from your statement, it seems that it is not needed to modify here. Is it right?